worker/dbaccessor/tracker.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package dbaccessor
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"gopkg.in/tomb.v2"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/database"
+)
+
+// TrackedDB defines the union of a TrackedDB and a worker.Worker interface.
+// This is local to the package, allowing for better testing of the underlying
+// trackerDB worker.
+type TrackedDB interface {
+	coredatabase.TrackedDB
+	worker.Worker
+}
+
+// TrackedDBWorkerOption is a function that configures a TrackedDBWorker.
+type TrackedDBWorkerOption func(*trackedDBWorker)
+
+// WithVerifyDBFunc sets the function used to verify the database connection.
+func WithVerifyDBFunc(f func(context.Context, *sql.DB) error) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.verifyDBFunc = f
+	}
+}
+
+// WithClock sets the clock used by the worker.
+func WithClock(clock clock.Clock) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.clock = clock
+	}
+}
+
+// WithLogger sets the logger used by the worker.
+func WithLogger(logger Logger) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.logger = logger
+	}
+}
+
+type trackedDBWorker struct {
+	tomb tomb.Tomb
+
+	dbApp     DBApp
+	namespace string
+
+	mutex sync.RWMutex
+	db    *sql.DB
+	err   error
+
+	clock  clock.Clock
+	logger Logger
+
+	verifyDBFunc func(context.Context, *sql.DB) error
+}
+
+// NewTrackedDBWorker creates a new TrackedDBWorker
+func NewTrackedDBWorker(dbApp DBApp, namespace string, opts ...TrackedDBWorkerOption) (TrackedDB, error) {
+	w := &trackedDBWorker{
+		dbApp:        dbApp,
+		namespace:    namespace,
+		clock:        clock.WallClock,
+		verifyDBFunc: defaultVerifyDBFunc,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	var err error
+	w.db, err = w.dbApp.Open(context.TODO(), w.namespace)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w.tomb.Go(w.loop)
+
+	return w, nil
+}
+
+// DB closes over a raw *sql.DB. Closing over the DB allows the late
+// realization of the database. Allowing retries of DB acquisition if there
+// is a failure that is non-retryable.
+func (w *trackedDBWorker) DB(fn func(*sql.DB) error) error {
+	w.mutex.RLock()
+	// We have a fatal error, the DB can not be accessed.
```

Comment:
If we are giving a fatal error, I prefer to do this comment inside the if, or make this a question "// Do we have a fatal error?"


Also, do I understand the purpose of the closure correctly. This is *in case* we want to support retries in the future, but otherwise it doesn't actually do anything currently that we couldn't do with a simple 'return self.db')

I suppose it also means the pattern for callers can't be to ever cache the db reference, since they can only access it from a method like this. (we *could* say that you should always call DoSomething(DB()), but it would always be tempting to keep that pointer.

Also, the sentence "Allowing retries" is not a complete sentence.